File: 01.py
from aoc import *

DAY = 1

s = get_input(DAY)
#s = get_example(DAY)

import collections
import math

ls = s.splitlines()


def part1():
    a = []
    b = []
    for l in ls:
        c, d = ints(l)
        a.append(c)
        b.append(d)

    a.sort()
    b.sort()

    return sum(abs(c - d) for c, d in zip(a, b))


def part2():
    a = []
    b = []
    for l in ls:
        c, d = ints(l)
        a.append(c)
        b.append(d)

    bb = collections.Counter(b)

    return sum(c * bb[c] for c in a)


# submit(DAY, part1(), 1)
# submit(DAY, part2(), 2)

File: kotlin/src/01.kt
import kotlin.time.measureTimedValue
import kotlin.math.abs

fun main() {
    val ls = readInput()

    val locationIDsA = mutableListOf<Int>()
    val locationIDsB = mutableListOf<Int>()

    for (l in ls) {
        val (a, b) = ints(l)
        locationIDsA.addLast(a)
        locationIDsB.addLast(b)
    }

    fun part1(): Int {
        return locationIDsA.sorted().zip(locationIDsB.sorted()) { a, b -> abs(a - b) }.sum()
    }

    fun part2(): Int {
        val locationIDsBCounter = locationIDsB.toCounter()
        return locationIDsA.sumOf { a -> a * locationIDsBCounter.getOrDefault(a, 0) }
    }

    println(measureTimedValue { part1() }.let { "${it.value} (${it.duration})" })
    println(measureTimedValue { part2() }.let { "${it.value} (${it.duration})" })
}
